server/src/main/java/org/elasticsearch/index/mapper/Mapper.java

```diff
@@ -24,7 +24,7 @@ public abstract class Mapper implements ToXContentFragment, Iterable<Mapper> {
 
     public abstract static class Builder {
 
-        protected final String name;
+        protected String name;
```

Comment:
this is odd. I do see why you need it though. You get a builder from the mapper, and want to build a new mapper with only the field name changed. One solution could be to introduce a new method that's specific to this scenario, but that would need to be added to every single mapper. Perhaps it could be returning the mapper directly and taking the new name as an argument.

One thing that bugs me is that the member is protected, and used directly in many places, while there is a public getter for it. It would be easier to track who uses the setter (say we introduce it) if the field was private.  Another aspect that is not great is that we make the field non final on Mapper.Builder, but in practice this is only ever needed in FieldMapper.Builder . 

I haven't come up with an appealing enough alternative but I think it would be good to rethink how we got this bit done.

Reply:
I think it's fine for a builder to have mutable state - that's what it is made for. Only when you build the actual object, you'll get an immutable instance. I like your idea of encapsulating the instance variable more. I could make it private and force using the getter instead. The setter can then also be package-private.

Fun fact, the field used to be non-final before https://github.com/elastic/elasticsearch/pull/70837

Reply:
I see what you mean. I just find it hard to trace why the name has to change, and it's counter intuitive that it has a name set that later has to change.

> I like your idea of encapsulating the instance variable more. I could make it private and force using the getter instead. The setter can then also be package-private.

Shall we make that change separately so we don't introduce additional changes to this PR?

Reply:
Sure, here's the PR: https://github.com/elastic/elasticsearch/pull/105648

Reply:
> I see what you mean. I just find it hard to trace why the name has to change, and it's counter intuitive that it has a name set that later has to change.

I get what you mean. But I don't think the effort of implementing a new method in every field mapper is worth it. As this is a builder after all, I don't think it's bad practice to make it mutable. With the setter being package-protected, we also don't expose the setter to everyone.

Reply:
agreed, that's a good trade-off.

Reply:
Cool, I've merged https://github.com/elastic/elasticsearch/pull/105648 which encapsulates `Mapper.Builder#name` and then merged main into this PR.